### PR TITLE
Implement a REST API for developers to manage CORS at the application level

### PR DESCRIPTION
**Pull request by ivantha**
### Proposed changes in this pull request
$subject.

Closes https://github.com/wso2/product-is/issues/8877

### When to merge
After https://github.com/wso2/carbon-identity-framework/pull/3029

### CODE REVIEW

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.common/src/main/java/org/wso2/carbon/identity/api/server/application/management/common/ApplicationManagementServiceHolder.java [hunk 1]
```diff
@@ -110,4 +106,15 @@ public void setTemplateManager(TemplateManager templateManager) {
 
         ApplicationManagementServiceHolder.templateManager = templateManager;
     }
+
+    public CORSManagementService getCorsManagementService() {
+
+        return corsManagementService;
+    }
+
+    public void setCorsManagementService(
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.common/src/main/java/org/wso2/carbon/identity/api/server/application/management/common/ApplicationManagementServiceHolder.java:115]: inline lines 115 and 116

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/inbound/oauth2/OAuthInboundFunctions.java [hunk 2]
```diff
@@ -69,30 +95,40 @@ public static InboundAuthenticationRequestConfig putOAuthInbound(ServiceProvider
...
+        } catch (IdentityOAuthAdminException | CORSManagementServiceException e) {
+            if (e instanceof IdentityOAuthAdminException) {
+                // If an IdentityOAuthAdminException exception is thrown after the CORS update, then the application
+                // update has failed. Therefore rollback the update on CORS origins.
+                try {
+                    ApplicationManagementServiceHolder.getInstance()
+                            .getCorsManagementService().setApplicationCORSOrigins(tenantDomain,
+                            String.valueOf(application.getApplicationID()), existingCORSOrigins);
+                } catch (CORSManagementServiceException corsManagementServiceException) {
+                    log.error("Failed to update the CORS origins on rollback.", e);
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/inbound/oauth2/OAuthInboundFunctions.java:111]: no need to log here.. handleException will log the exception IINM

File: components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.common/src/main/java/org/wso2/carbon/identity/api/server/configs/common/Constants.java [hunk 3]
```diff
@@ -46,13 +62,49 @@ private Constants() {
...
+                "error while retrieving the server configs."),
+
+        //TODO: What should be the error prefix?
+        /**
+         * Unable to retrieve CORS origins.
+         */
+        ERROR_CODE_CORS_CONFIG_RETRIEVE("RST-65001",
+                "Unable to retrieve CORS configuration.",
+                "Server encountered an error while retrieving the CORS configuration."),
+
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.common/src/main/java/org/wso2/carbon/identity/api/server/configs/common/Constants.java:87]: shall we remove unncessary new lines to be consistent of previously defined error enums

File: components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.v1/src/main/java/org/wso2/carbon/identity/api/server/configs/v1/function/CORSConfigurationToCORSConfig.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
+package org.wso2.carbon.identity.api.server.configs.v1.function;
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.v1/src/main/java/org/wso2/carbon/identity/api/server/configs/v1/function/CORSConfigurationToCORSConfig.java:1]: add license header

File: components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.v1/src/main/java/org/wso2/carbon/identity/api/server/configs/v1/function/CORSOriginToCORSOriginGetObject.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
+package org.wso2.carbon.identity.api.server.configs.v1.function;
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.v1/src/main/java/org/wso2/carbon/identity/api/server/configs/v1/function/CORSOriginToCORSOriginGetObject.java:1]: add license header

File: components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.v1/src/main/resources/configs.yaml [hunk 6]
```diff
@@ -14,7 +14,7 @@ security:
   - OAuth2: []
   - BasicAuth: []
 paths:
-  /configs:
+  /configs/:
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.v1/src/main/resources/configs.yaml:17]: why was this changed?

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/inbound/oauth2/OAuthInboundFunctions.java [hunk 7]
```diff
@@ -69,30 +95,39 @@ public static InboundAuthenticationRequestConfig putOAuthInbound(ServiceProvider
...
+                // Create a new application.
                 return createOAuthInbound(application.getApplicationName(), oidcConfigModel);
             }
 
-        } catch (IdentityOAuthAdminException e) {
-            throw handleOAuthException(e);
+        } catch (IdentityOAuthAdminException | CORSManagementServiceException e) {
+            if (e instanceof IdentityOAuthAdminException) {
+                // If an IdentityOAuthAdminException exception is thrown after the CORS update, then the application
+                // update has failed. Therefore rollback the update on CORS origins.
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/inbound/oauth2/OAuthInboundFunctions.java:105]: better to use a multi-line comment here.

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/inbound/oauth2/OAuthInboundFunctions.java [hunk 8]
```diff
@@ -131,22 +169,68 @@ public static OpenIDConnectConfiguration getOAuthConfiguration(InboundAuthentica
...
-            return new OAuthConsumerAppToApiModel().apply(oauthApp);
 
-        } catch (IdentityOAuthAdminException e) {
+            OpenIDConnectConfiguration openIDConnectConfiguration = new OAuthConsumerAppToApiModel().apply(oauthApp);
+
+            // Set CORS origins as allowed domains.
+            String tenantDomain = ContextLoader.getTenantDomainFromContext();
+            String applicationResourceId = ApplicationManagementServiceHolder.getInstance()
+                    .getApplicationManagementService()
+                    .getServiceProviderByClientId(clientId, "oauth2", tenantDomain).getApplicationResourceId();
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/inbound/oauth2/OAuthInboundFunctions.java:179]: Shall we add a constant for "oauth2"?
